bgpd: conditional advertisement

maduri111 opened the pull request:
Implemented as per the feature description given in the source link.

Descriprion:
The BGP conditional advertisement feature uses the non-exist-map or exist-map
and the advertise-map keywords of the neighbor advertise-map command in order
to track routes by the route prefix.

non-exist-map :
If a route prefix is not present in output of the non-exist-map command, then
the route specified by the advertise-map command is announced.

exist-map :
If a route prefix is present in output of the exist-map command, then the route
specified by the advertise-map command is announced.

The conditional BGP announcements are sent in addition to the normal
announcements that a BGP router sends to its peers.

The conditional advertisement process is triggered by the BGP scanner process,
which runs every 60 seconds. This means that the maximum time for the conditional
advertisement to take effect is 60 seconds. The conditional advertisement can take
effect sooner, depending on when the tracked route is removed from the BGP table
and when the next instance of the BGP scanner occurs.

Sample Configuration on DUT
---------------------------
Router2# show running-config
Building configuration...

Current configuration:
!
frr version 7.6-dev-MyOwnFRRVersion
frr defaults traditional
hostname router
log file /var/log/frr/bgpd.log
log syslog informational
hostname Router2
service integrated-vtysh-config
!
debug bgp updates in
debug bgp updates out
!
debug route-map
!
ip route 130.130.0.0/16 blackhole
!
interface enp0s9
 ip address 10.10.10.2/24
!
interface enp0s10
 ip address 10.10.20.2/24
!
interface lo
 ip address 128.16.16.1/24
!
router bgp 2
 bgp log-neighbor-changes
 no bgp ebgp-requires-policy
 neighbor 10.10.10.1 remote-as 1
 neighbor 10.10.20.3 remote-as 3
 !
 address-family ipv4 unicast
  network 128.16.16.0/24
  network 130.130.0.0/16
  neighbor 10.10.10.1 soft-reconfiguration inbound
  neighbor 10.10.10.1 advertise-map ADVERTISE non-exist-map CONDITION
  neighbor 10.10.20.3 soft-reconfiguration inbound
 exit-address-family
!
access-list 60 seq 5 permit 128.16.16.0/24 exact-match
access-list 65 seq 5 permit 192.168.50.0/24 exact-match
access-list CONDITION seq 5 permit 192.168.50.0/24
access-list ADVERTISE seq 5 permit 128.16.16.0/24
!
route-map ADVERTISE permit 10
 match ip address ADVERTISE
!
route-map CONDITION permit 10
 match ip address CONDITION
!
line vty
!
end
Router2#

Sample output
-------------
Router2# show ip bgp neighbors 10.10.10.1
BGP neighbor is 10.10.10.1, remote AS 1, local AS 2, external link

!--- Output suppressed.

 For address family: IPv4 Unicast
  Update group 1, subgroup 1
  Packet Queue length 0
  Inbound soft reconfiguration allowed
  Community attribute sent to this neighbor(all)
  Condition NON_EXIST, Condition-map *CONDITION, Advertise-map *ADVERTISE, status: Withdraw
  1 accepted prefixes

  Connections established 1; dropped 0
  Last reset 00:21:05,  Waiting for peer OPEN

!--- Output suppressed.

Router2#

Router2# show ip bgp
BGP table version is 6, local router ID is 128.16.16.1, vrf id 0
Default local pref 100, local AS 2
Status codes:  s suppressed, d damped, h history, * valid, > best, = multipath,
               i internal, r RIB-failure, S Stale, R Removed
Nexthop codes: @NNN nexthop's vrf id, < announce-nh-self
Origin codes:  i - IGP, e - EGP, ? - incomplete

   Network          Next Hop            Metric LocPrf Weight Path
*> 128.16.16.0/24   0.0.0.0                  0         32768 i
*> 130.130.0.0/16   0.0.0.0                  0         32768 i
*> 192.168.50.0/24  10.10.20.3               0             0 3 i
*> 200.200.200.0/24 10.10.10.1               0             0 1 i

Displayed  4 routes and 4 total paths
Router2#

Here 128.16.16.0/24(prefix in advertise-map) is withdrawn by conditional
advertisement scanner as the prefix(192.168.50.0/24) specified by
non-exist-map is present in BGP table.

Router2# show ip bgp neighbors 10.10.10.1 advertised-routes
BGP table version is 6, local router ID is 128.16.16.1, vrf id 0
Default local pref 100, local AS 2
Status codes:  s suppressed, d damped, h history, * valid, > best, = multipath,
               i internal, r RIB-failure, S Stale, R Removed
Nexthop codes: @NNN nexthop's vrf id, < announce-nh-self
Origin codes:  i - IGP, e - EGP, ? - incomplete

   Network          Next Hop            Metric LocPrf Weight Path
*> 130.130.0.0/16   0.0.0.0                  0         32768 i
*> 192.168.50.0/24  0.0.0.0                                0 3 i
*> 200.200.200.0/24 0.0.0.0                                0 1 i

Total number of prefixes 3
Router2#

After Removing 192.168.50.0/24(prefix present in non-exist-map), prefix present in
advertise-map(128.16.16.0/24) is advertised

Router2# show ip bgp
BGP table version is 7, local router ID is 128.16.16.1, vrf id 0
Default local pref 100, local AS 2
Status codes:  s suppressed, d damped, h history, * valid, > best, = multipath,
               i internal, r RIB-failure, S Stale, R Removed
Nexthop codes: @NNN nexthop's vrf id, < announce-nh-self
Origin codes:  i - IGP, e - EGP, ? - incomplete

   Network          Next Hop            Metric LocPrf Weight Path
*> 128.16.16.0/24   0.0.0.0                  0         32768 i
*> 130.130.0.0/16   0.0.0.0                  0         32768 i
*> 200.200.200.0/24 10.10.10.1               0             0 1 i

Displayed  3 routes and 3 total paths
Router2#
Router2#
Router2# show ip bgp neighbors 10.10.10.1 advertised-routes
BGP table version is 7, local router ID is 128.16.16.1, vrf id 0
Default local pref 100, local AS 2
Status codes:  s suppressed, d damped, h history, * valid, > best, = multipath,
               i internal, r RIB-failure, S Stale, R Removed
Nexthop codes: @NNN nexthop's vrf id, < announce-nh-self
Origin codes:  i - IGP, e - EGP, ? - incomplete

   Network          Next Hop            Metric LocPrf Weight Path
*> 128.16.16.0/24   0.0.0.0                  0         32768 i
*> 130.130.0.0/16   0.0.0.0                  0         32768 i
*> 200.200.200.0/24 0.0.0.0                                0 1 i

Total number of prefixes 3
Router2#
Router2#
Router2#

Signed-off-by: Madhuri Kuruganti <k.madhuri@samsung.com>

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/085a125a70429c3b678733b09bfad37d/raw/ea961d3cbbf3168505e8a61dbd45189e35833780/cr_7189_1601249031.diff | git apply
```

</p>

<p>

```diff
diff --git a/bgpd/bgp_route.h b/bgpd/bgp_route.h
index b1edf6ef8..f82759b11 100644
--- a/bgpd/bgp_route.h
+++ b/bgpd/bgp_route.h
@@ -430,13 +430,13 @@ struct bgp_aggregate {
 #define UNSUPPRESS_MAP_NAME(F)  ((F)->usmap.name)
 #define UNSUPPRESS_MAP(F)       ((F)->usmap.map)
 
-#define ADVERTISE_MAP_NAME(F)	((F)->advmap.aname)
-#define ADVERTISE_MAP(F)	((F)->advmap.amap)
+#define ADVERTISE_MAP_NAME(F) ((F)->advmap.aname)
+#define ADVERTISE_MAP(F) ((F)->advmap.amap)
 
-#define ADVERTISE_CONDITION(F)	((F)->advmap.condition)
+#define ADVERTISE_CONDITION(F) ((F)->advmap.condition)
 
-#define CONDITION_MAP_NAME(F)	((F)->advmap.cname)
-#define CONDITION_MAP(F)	((F)->advmap.cmap)
+#define CONDITION_MAP_NAME(F) ((F)->advmap.cname)
+#define CONDITION_MAP(F) ((F)->advmap.cmap)
 
 /* path PREFIX (addpath rxid NUMBER) */
 #define PATH_ADDPATH_STR_BUFFER PREFIX2STR_BUFFER + 32
diff --git a/bgpd/bgp_vty.c b/bgpd/bgp_vty.c
index aff1c5a2b..fbaabf5ce 100644
--- a/bgpd/bgp_vty.c
+++ b/bgpd/bgp_vty.c
@@ -6844,17 +6844,15 @@ static int peer_advertise_map_unset_vty(struct vty *vty, const char *ip_str,
 	return bgp_vty_return(vty, ret);
 }
 
-DEFUN (neighbor_advertise_map,
-       neighbor_advertise_map_cmd,
-       "neighbor <A.B.C.D|X:X::X:X|WORD> advertise-map WORD <exist-map WORD|non-exist-map WORD>",
-       NEIGHBOR_STR
-       NEIGHBOR_ADDR_STR2
-       "Route-map to conditionally advertise routes\n"
-       "Name of advertise map\n"
-       "Advertise routes only if prefixes in exist-map are installed in BGP table\n"
-       "Name of the exist map\n"
-       "Advertise routes only if prefixes in non-exist-map are not installed in BGP table\n"
-       "Name of the non exist map\n")
+DEFUN(neighbor_advertise_map, neighbor_advertise_map_cmd,
+      "neighbor <A.B.C.D|X:X::X:X|WORD> advertise-map WORD <exist-map WORD|non-exist-map WORD>",
+      NEIGHBOR_STR NEIGHBOR_ADDR_STR2
+      "Route-map to conditionally advertise routes\n"
+      "Name of advertise map\n"
+      "Advertise routes only if prefixes in exist-map are installed in BGP table\n"
+      "Name of the exist map\n"
+      "Advertise routes only if prefixes in non-exist-map are not installed in BGP table\n"
+      "Name of the non exist map\n")
 {
 	int idx = 0;
 	int idx_peer = 1;
@@ -6871,28 +6869,26 @@ DEFUN (neighbor_advertise_map,
 		argv[idx_condition_word]->arg);
 }
 
-ALIAS_HIDDEN(neighbor_advertise_map, neighbor_advertise_map_hidden_cmd,
-	     "neighbor <A.B.C.D|X:X::X:X|WORD> advertise-map WORD <exist-map WORD|non-exist-map WORD>",
-	     NEIGHBOR_STR NEIGHBOR_ADDR_STR2
-	     "Route-map to conditionally advertise routes\n"
-	     "Name of advertise map\n"
-	     "Advertise routes only if prefixes in exist-map are installed in BGP table\n"
-	     "Name of the exist map\n"
-	     "Advertise routes only if prefixes in non-exist-map are not installed in BGP table\n"
-	     "Name of the non exist map\n")
-
-DEFUN (no_neighbor_advertise_map,
-       no_neighbor_advertise_map_cmd,
-       "no neighbor <A.B.C.D|X:X::X:X|WORD> advertise-map WORD <exist-map WORD|non-exist-map WORD>",
-       NO_STR
-       NEIGHBOR_STR
-       NEIGHBOR_ADDR_STR2
-       "Route-map to conditionally advertise routes\n"
-       "Name of advertise map\n"
-       "Advertise routes only if prefixes in exist-map are installed in BGP table\n"
-       "Name of the exist map\n"
-       "Advertise routes only if prefixes in non-exist-map are not installed in BGP table\n"
-       "Name of the non exist map\n")
+ALIAS_HIDDEN(
+	neighbor_advertise_map, neighbor_advertise_map_hidden_cmd,
+	"neighbor <A.B.C.D|X:X::X:X|WORD> advertise-map WORD <exist-map WORD|non-exist-map WORD>",
+	NEIGHBOR_STR NEIGHBOR_ADDR_STR2
+	"Route-map to conditionally advertise routes\n"
+	"Name of advertise map\n"
+	"Advertise routes only if prefixes in exist-map are installed in BGP table\n"
+	"Name of the exist map\n"
+	"Advertise routes only if prefixes in non-exist-map are not installed in BGP table\n"
+	"Name of the non exist map\n")
+
+DEFUN(no_neighbor_advertise_map, no_neighbor_advertise_map_cmd,
+      "no neighbor <A.B.C.D|X:X::X:X|WORD> advertise-map WORD <exist-map WORD|non-exist-map WORD>",
+      NO_STR NEIGHBOR_STR NEIGHBOR_ADDR_STR2
+      "Route-map to conditionally advertise routes\n"
+      "Name of advertise map\n"
+      "Advertise routes only if prefixes in exist-map are installed in BGP table\n"
+      "Name of the exist map\n"
+      "Advertise routes only if prefixes in non-exist-map are not installed in BGP table\n"
+      "Name of the non exist map\n")
 {
 	int idx = 0;
 	int idx_peer = 2;
@@ -6909,15 +6905,16 @@ DEFUN (no_neighbor_advertise_map,
 		argv[idx_condition_word]->arg);
 }
 
-ALIAS_HIDDEN(no_neighbor_advertise_map, no_neighbor_advertise_map_hidden_cmd,
-	     "no neighbor <A.B.C.D|X:X::X:X|WORD> advertise-map WORD <exist-map WORD|non-exist-map WORD>",
-	     NO_STR NEIGHBOR_STR NEIGHBOR_ADDR_STR2
-	     "Route-map to conditionally advertise routes\n"
-	     "Name of advertise map\n"
-	     "Advertise routes only if prefixes in exist-map are installed in BGP table\n"
-	     "Name of the exist map\n"
-	     "Advertise routes only if prefixes in non-exist-map are not installed in BGP table\n"
-	     "Name of the non exist map\n")
+ALIAS_HIDDEN(
+	no_neighbor_advertise_map, no_neighbor_advertise_map_hidden_cmd,
+	"no neighbor <A.B.C.D|X:X::X:X|WORD> advertise-map WORD <exist-map WORD|non-exist-map WORD>",
+	NO_STR NEIGHBOR_STR NEIGHBOR_ADDR_STR2
+	"Route-map to conditionally advertise routes\n"
+	"Name of advertise map\n"
+	"Advertise routes only if prefixes in exist-map are installed in BGP table\n"
+	"Name of the exist map\n"
+	"Advertise routes only if prefixes in non-exist-map are not installed in BGP table\n"
+	"Name of the non exist map\n")
 
 /* Set route-map to the peer. */
 static int peer_route_map_set_vty(struct vty *vty, const char *ip_str,
diff --git a/bgpd/bgpd.c b/bgpd/bgpd.c
index 1264d0dfb..652c6809a 100644
--- a/bgpd/bgpd.c
+++ b/bgpd/bgpd.c
@@ -4908,7 +4908,7 @@ void peer_port_unset(struct peer *peer)
  * initiates inbound or outbound processing as needed.
  */
 void peer_on_policy_change(struct peer *peer, afi_t afi, safi_t safi,
-				  int outbound)
+			   int outbound)
 {
 	if (outbound) {
 		update_group_adjust_peer(peer_af_find(peer, afi, safi));
diff --git a/bgpd/bgpd.h b/bgpd/bgpd.h
index 5cff254b3..4ef2f841e 100644
--- a/bgpd/bgpd.h
+++ b/bgpd/bgpd.h
@@ -763,8 +763,8 @@ struct bgp_nexthop {
 #define BGP_GTSM_HOPS_CONNECTED 1
 
 /* Advertise map */
-#define CONDITION_NON_EXIST	0
-#define CONDITION_EXIST     	1
+#define CONDITION_NON_EXIST 0
+#define CONDITION_EXIST 1
 
 #include "filter.h"
 
@@ -1375,7 +1375,7 @@ struct peer {
 #define PEER_FT_PREFIX_LIST           (1U << 2) /* prefix-list */
 #define PEER_FT_ROUTE_MAP             (1U << 3) /* route-map */
 #define PEER_FT_UNSUPPRESS_MAP        (1U << 4) /* unsuppress-map */
-#define PEER_FT_ADVERTISE_MAP         (1U << 5) /* advertise-map */
+#define PEER_FT_ADVERTISE_MAP (1U << 5)		/* advertise-map */
 
 	/* ORF Prefix-list */
 	struct prefix_list *orf_plist[AFI_MAX][SAFI_MAX];
@@ -2217,6 +2217,6 @@ DECLARE_HOOK(peer_status_changed, (struct peer * peer), (peer))
 void peer_nsf_stop(struct peer *peer);
 
 extern void peer_on_policy_change(struct peer *peer, afi_t afi, safi_t safi,
-                                  int outbound);
+				  int outbound);
 
 #endif /* _QUAGGA_BGPD_H */

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

rzalamena commented on bgpd/bgp_conditional_adv.h:
```diff
@@ -0,0 +1,130 @@
+
+	/* Doesn't make sense to configure advertise
+	 * or condition map in deny/any clause. */
+	if (!rmap->head->type == RMAP_PERMIT)
```
OpenBSD compiler is rightfully complaining here: do we want to negate `rmap->head->type` or the condition `rmap->head->type == RMAP_PERMIT`?

Please use parentheses to disambiguate.

maduri111 replied:
its if (rmap->head->type != RMAP_PERMIT)

maduri111 replied:
this function is removed after added code to support other match rules to conditional advertisement

maduri111 replied:
used the existing route_map_apply()

rzalamena reviewed (commented):
Code looks good, but the PR itself lacks some items:

1. Documentation
2. Tests

I added some comments to help clarifying some parts.

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+					continue;
+			}
+
+			/* Derive conditional advertisement staus from condition
```
type: status

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+				filter->advmap.status = false;
+			else {
+				/* no change in status. we only advertise
+				 * previosly withdrawn routes if needed. */
```
typo: previously

ton31337 commented on bgpd/bgpd.h:
```diff
@@ -681,6 +681,10 @@ struct bgp {
 
+	/* BGP Conditional advertisement */
+	int condition_filter_count;
+	struct thread *t_condition_check;
```
Shouldn't be this per AFI/SAFI?

maduri111 replied:
No Only one timer thread for polling. 
timer thread will be created/started when advertise-map is created on first neighbor
and will be canceled when the advertise-map is removed on the last neighbor.

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+
+		/* If a match command refers to several objects in one
+		 * command either of them should match (i.e logical OR) */
+		FOREACH_ACCESS_LIST_FILTER(alist, alist_filter) {
```
`route_map_apply()` wouldn't be a good candidate here?

maduri111 replied:
Sorry, route-map-apply() cannot be used as it is. because here the lookup is reverse.
Route-map : 'For each prefix in BGP table', we check for the best match in route-map & 
                      if there is no best match, route-map action is to deny the prefix and withdraw the route. 
advertise-map : 'For each prefix in route-map match filter' if there is a match or no match in BGP table, 
                          action will be taken based on prefix match, condition(exist/non-exist) and previous state.
                          and if there is a config change or not. this is to avoid sending duplicate advertisements.
 Also every 60 seconds we need to scan for table entries. 
So to avoid additional look ups and unnecessary checks written new functions which are necessary to be checked.

ton31337 reviewed (changes requested):
The code more or less looks good, just some questions from me. Docs+topotests are missing. CI warnings must be addressed as well.

maduri111 commented on bgpd/bgpd.c:
```diff
@@ -6527,6 +6528,190 @@ int peer_unsuppress_map_unset(struct peer *peer, afi_t afi, safi_t safi)
+	 */
+	for (ALL_LIST_ELEMENTS(peer->group->peer, node, nnode, member)) {
+		/* Skip peers with overridden configuration. */
+		if (CHECK_FLAG(member->filter_override[afi][safi][1],
```
will replace 1 with RMAP_OUT.

maduri111 commented on bgpd/bgpd.c:
```diff
@@ -6527,6 +6528,190 @@ int peer_unsuppress_map_unset(struct peer *peer, afi_t afi, safi_t safi)
+	struct listnode *node, *nnode;
+
+	/* Unset override-flag unconditionally. */
+	UNSET_FLAG(peer->filter_override[afi][safi][0], PEER_FT_ADVERTISE_MAP);
```
Sorry it should be 1 ie RMAP_OUT. its direction. for this feature it should be always RMAP_OUT.

maduri111 replied:
addressed this comment

maduri111 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+		return -1;
+
+	/* Last filter removed. So cancel contional routes polling thread. */
+	if (bgp->t_condition_check)
```
Sure.

maduri111 replied:
addressed this comment

maduri111 commented on bgpd/bgp_conditional_adv.h:
```diff
@@ -0,0 +1,130 @@
+		return AFI_MAX;
+}
+
+static inline bool advertise_dest_routes(struct update_subgroup *subgrp,
```
> 
> 
> This function seems a bit big to be inlined, is there any specific reason?

We are calling this function for each bgp_dest node in the table.
if the Table size is big there could be more number of function calls thats why. and we poll the table every 60seconds once. If you feel its ok to remove inline. will do that.

maduri111 replied:
> 
> 
> I have another one question regarding this function: can't `subgroup_announce_table` be adopted for all the changes we do here? I mean you introduced a few functions which basically do almost the same with some additional conditions.

for some cases subgroup_announce_table() is used. but not for all to avoid unnecessary checks and look-up
example route-map attached to advertise-map is removed but previously some routes were withdrawn.
in that case subgroup_announce_table() is used to send normal update.

maduri111 replied:
this function is removed after adding code to support other match rules.

maduri111 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+
+	bgp = THREAD_ARG(t);
+
+	if (!bgp)
```
Done

maduri111 commented on bgpd/bgpd.c:
```diff
@@ -6527,6 +6528,190 @@ int peer_unsuppress_map_unset(struct peer *peer, afi_t afi, safi_t safi)
+	 */
+	for (ALL_LIST_ELEMENTS(peer->group->peer, node, nnode, member)) {
+		/* Skip peers with overridden configuration. */
+		if (CHECK_FLAG(member->filter_override[afi][safi][0],
```
same here it should be RMAP_OUT

maduri111 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+	struct filter_cisco *cfilter = NULL;
+	struct filter_zebra *zfilter = NULL;
+
+	if (!table || !filter)
```
Sure. will check in other places as well and remove additional checks.

maduri111 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+	return 0;
+}
+
+int bgp_conditional_adv_enable(struct peer *peer, afi_t afi, safi_t safi)
```
Yes. will change enable and disable functions return type to void

maduri111 commented on bgpd/bgp_conditional_adv.h:
```diff
@@ -0,0 +1,130 @@
+	return true;
+}
+
+static inline afi_t get_afi_from_str(const char *str)
```
ok. Shall I change the function name to get_afi_from_match_rule()?

maduri111 commented on bgpd/bgp_routemap.c:
```diff
@@ -3695,6 +3695,18 @@ static void bgp_route_map_process_peer(const char *rmap_name,
+	if (filter->advmap.aname
+	    && (strcmp(rmap_name, filter->advmap.aname) == 0)) {
+		filter->advmap.amap = map;
+		peer->advmap_info[afi][safi].config_change = true;
```
This variable will be set when there is configuration change. and is used to decide its config change/ change in table/no change compared to previous scan. For example 
1. condition change => exist-map to non-exist-map or vice versa
2. advertise-map <WORD> => route-map attached to advertise-map is changed
3. exist-map <WORD> or non-exist-map => route-map attached to exist-map or non-exist-map changed
4. route-map (attached to advertise-map/exist-map/non-exist-map) removed/modified
5. match rule removed/modified/added
6. acl is removed/added/modified

maduri111 commented on bgpd/bgpd.h:
```diff
@@ -791,6 +799,19 @@ struct bgp_filter {
+		char *cname;
+		struct route_map *cmap;
+
+		bool status;
```
in show ip bgp neighbors A.B.C.D command the output shows that as 'status' hence used it as it is.
Condition-map NON-EXIST, Advertise-map ADVERTISE, status: Advertise

Shall I change the variable name to advertise ? or keep it as status ?

ton31337 replied:
Yes, but frankly speaking status can't be just true or false :) enums in such cases are handier, IMO.

maduri111 commented on bgpd/bgpd.h:
```diff
@@ -791,6 +799,19 @@ struct bgp_filter {
+		char *aname;
+		struct route_map *amap;
+
+		bool condition;
```
here also, initially i used the variable name as exist but in show command it shows as Condition-map 'EXIST or NON-EXIST'.
please let me know if you still want to change this variable name

ton31337 replied:
But looking at the code confuses a bit what does it actually mean.

maduri111 replied:
ADVERTISE_MAP and CONDITION_MAP are user defined route-map names.

neighbor 10.10.10.1 advertise-map ADVERTISE_MAP exist-map CONDITION_MAP
in this command, condition variable will be set to CONDITION_EXIST
neighbor 10.10.10.1 advertise-map ADVERTISE_MAP non-exist-map CONDITION_MAP
in this command condition variable will be set to CONDITION_NONEXIST

if condition == CONDITION_EXIST and 
CONDITION_MAP's prefix(es) are in BGP table then --> ADVERTISE_MAP's prefixes in BGP table will be advertised.
CONDITION_MAP's prefix(es) are not in BGP table then --> ADVERTISE_MAP's prefixes in BGP table will be withdrawn.

if condition == CONDITION_NONEXIST and 
CONDITION_MAP's prefix(es) are in BGP table then --> ADVERTISE_MAP's prefixes in BGP table will be withdrawn.
CONDITION_MAP's prefix(es) are not in BGP table then --> ADVERTISE_MAP's prefixes in BGP table will be advertised.

donaldsharp commented:
This is 1k of change in 1 commit.  Is there any way that this commit could be broken up into smaller logical units?  I am also going to echo the doc + topotest changes.  We must have them before this code change can be accepted.

maduri111 commented:
> 
> 
> The code more or less looks good, just some questions from me. Docs+topotests are missing. CI warnings must be addressed as well.

Sure will do that.

maduri111 commented:
> 
> 
> This is 1k of change in 1 commit. Is there any way that this commit could be broken up into smaller logical units? I am also going to echo the doc + topotest changes. We must have them before this code change can be accepted.

Ok. Will try to send asap. Thanks!

maduri111 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+}
+
+enum route_map_cmd_result_t
+bgp_check_rmap_prefixes_in_bgp_table(struct bgp_table *table,
```
we want to use the ACL filter's prefix as the key to look for match in BGP table right. so cannot use access_list_apply()

maduri111 replied:
reused the existing functions.

maduri111 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+
+/* We just need bgp_dest node matches with filter prefix. So no need to
+ * traverse each path here. */
+struct bgp_dest *bgp_dest_matches_filter_prefix(struct bgp_table *table,
```
same here as well. 
we want to use the ACL filter's prefix as the key to look for match in BGP table right.

maduri111 commented on doc/user/bgp.rst:
```diff
@@ -2564,6 +2564,182 @@ This makes it possible to separate not only layer 3 networks like VRF-lite netwo
+
+.. index:: [no] neighbor A.B.C.D advertise-map NAME non-exist-map NAME
+.. clicmd:: [no] neighbor A.B.C.D advertise-map NAME non-exist-map NAME
+
```
.. index:: [no] neighbor A.B.C.D advertise-map NAME [exist-map|non-exist-map] NAME
.. clicmd:: [no] neighbor A.B.C.D advertise-map NAME [exist-map|non-exist-map] NAME

maduri111 replied:
Sorry, will correct it in next commit while uploading topotest.

maduri111 commented:
> 
> 
> This is 1k of change in 1 commit. Is there any way that this commit could be broken up into smaller logical units? I am also going to echo the doc + topotest changes. We must have them before this code change can be accepted.

@donaldsharp 
To split the patch into smaller logical units, Do I need to revert the patch and commit the changes in smaller chunks ?
Please let me know the correct way.

mwinter-osr commented:
@maduri111 To fix the errors in the topotests, please rebase to latest master. A few fixes just got merged into master which should resolve these issues

maduri111 commented:
> 
> 
> @maduri111 To fix the errors in the topotests, please rebase to latest master. A few fixes just got merged into master which should resolve these issues

Done. Thank you

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,451 @@
+		/* If a match command refers to several objects in one
+		 * command either of them should match (i.e logical OR)
+		 */
+		FOREACH_ACCESS_LIST_FILTER(alist, alist_filter) {
```
prefix-lists can be used in route-maps for exists-map/non-exists-map? Or only access-lists?

maduri111 replied:
Sorry I have added only access-list for now, will add prefix-list as well.

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,451 @@
+
+			filter = &peer->filter[afi][safi];
+
+			if ((!filter->advmap.aname) || (!filter->advmap.cname)
```
can we reduce `()` count?

maduri111 replied:
sure will address this comment

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,451 @@
+{
+	struct bgp *bgp = peer->bgp;
+
+	assert(bgp);
```
can this be NULL?

rzalamena replied:
probably not, but in case it is it will `assert` (because it is unexpected). I don't see harm in this and it will avoid coverity issue (if any).

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,451 @@
+	 * it is configured on more than one neighbor(AFI/SAFI).
+	 * So there's nothing to do except decrementing the counter.
+	 */
+	if (--bgp->condition_filter_count != 0)
```
can this be below 0? if not, maybe uint would be easier here to deal with?

maduri111 replied:
already changed it to uint32_t condition_filter_count;

ton31337 replied:
yep, sorry, this was an old pending comment :)

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,451 @@
+		return;
+
+	/* Last filter removed. So cancel conditional routes polling thread. */
+	THREAD_OFF(bgp->t_condition_check);
```
double-checking: Is this a global thread or should be per-peer?

maduri111 replied:
its global thread. not per peer. 
when advertise-map is configured on first peer it will be created and deleted when advertise-map on last peer is removed.

ton31337 replied:
and what happens if I have 100 peers each with its own advertise-map?

maduri111 replied:
thread will be deleted/OFF only after removing the advertise-map config on 100th peer.

ton31337 replied:
I'm not sure if this is the right way. Does it mean that with N peers it would take N*CONDITIONAL_ROUTES_POLL_TIME to complete?

maduri111 replied:
No. Always it will be 60 seconds, irrespective of the number of advertise-maps configured per BGP process.
when scanning process is called we will check if peer[afi][safi] has advertise-map configurations and process the pending event on each peer (AFI/SAFI).

ton31337 commented on bgpd/bgp_route.c:
```diff
@@ -1950,7 +1951,9 @@ bool subgroup_announce_check(struct bgp_dest *dest, struct bgp_path_info *pi,
 
 	/* Route map & unsuppress-map apply. */
-	if (ROUTE_MAP_OUT_NAME(filter) || (pi->extra && pi->extra->suppress)) {
+	if ((!skip_rmap_check)
```
no additional `()` preferred

maduri111 replied:
ok.

ton31337 commented on bgpd/bgpd.c:
```diff
@@ -6573,6 +6574,192 @@ int peer_unsuppress_map_unset(struct peer *peer, afi_t afi, safi_t safi)
+		SET_FLAG(peer->filter_override[afi][safi][RMAP_OUT],
+			 PEER_FT_ADVERTISE_MAP);
+
+		/* Hold peer_on_policy_change() until timer thread is called */
```
something is missing here or lonely comment? :)

maduri111 replied:
no just a comment. :)

ton31337 commented on bgpd/bgpd.h:
```diff
@@ -758,6 +762,10 @@ struct bgp_nexthop {
 #define BGP_GTSM_HOPS_CONNECTED 1
 
+/* Advertise map */
+#define CONDITION_NON_EXIST	0
```
since this is only bool, we should use true/false instead

maduri111 replied:
sure. will change it

ton31337 reviewed (commented):
by the way, is it possible to use a condition like: Send advertise-map only if match community 123:123?

maduri111 commented:
> 
> 
> by the way, is it possible to use a condition like: Send advertise-map only if match community 123:123?

I just checked. cisco supports match prefix-list and match community also. Will add the code for that as well and send the changes. Thanks.

maduri111 commented:
> 
> 
> by the way, is it possible to use a condition like: Send advertise-map only if match community 123:123?

@ton31337 
Modified code to support other match rules and tested with prefix-list, community. 
Removed new functions and reused the existing one. 
Please find the sample config and output in commit message.

ton31337 commented on bgpd/bgp_packet.c:
```diff
@@ -1782,6 +1782,9 @@ static int bgp_update_receive(struct peer *peer, bgp_size_t size)
 	peer->update_time = bgp_clock();
 
+	/* Notify BGP Conditional advertisement scanner process */
+	peer->advmap_table_change = true;
```
where is this used?

ton31337 commented on bgpd/bgp_route.c:
```diff
@@ -1990,6 +1993,27 @@ bool subgroup_announce_check(struct bgp_dest *dest, struct bgp_path_info *pi,
+					if (!CHECK_FLAG(peer->flags,
+							PEER_FLAG_CONFIG_NODE))
+						continue;
+					if (strcmp(peer->host, temp_peer->host)
```
I think this should be handled per subgroup and not per peer.

maduri111 replied:
advertise-map is per peer context so scanning happens per peer.
And this check is needed when we configure route-map outbound along with advertise-map and routes matching with condition-map/advertise-map are filtered out by route-map.
Configuring Route-map out and advertise-map is a valid scenario ?

ton31337 replied:
> Configuring Route-map out and advertise-map is a valid scenario ?

I think it is because you can have `rmap1` permitting 192.168.0.0/24, 192.168.1.0/24, and `advrmap1` advertising 192.168.1.0/24 only under conditions.

ton31337 replied:
Can you create such a setup for topotest?

ton31337 replied:
By the way, isn't enough to compare `peer == temp_peer`? or using `peer_cmp()` to make sure they are not in the same group, etc.

maduri111 replied:
addressed this comment. 
used peer == temp_peer. added test cases in topotest.

ton31337 commented on bgpd/bgp_vty.c:
```diff
@@ -7047,6 +7047,132 @@ ALIAS_HIDDEN(no_neighbor_filter_list, no_neighbor_filter_list_hidden_cmd,
+	if (!peer)
+		return ret;
+
+	condition_map = route_map_lookup_warn_noexist(vty, condition_str);
```
can we assign this directly above? `struct route_map *advertise_map = route_map_lookup_warn_noexist(vty, condition_str);`, etc.

maduri111 replied:
we have null check for peer before route_map_lookup_warn_noexist(). if peer is null its waste of look up right ?

ton31337 commented on bgpd/bgp_vty.c:
```diff
@@ -7047,6 +7047,132 @@ ALIAS_HIDDEN(no_neighbor_filter_list, no_neighbor_filter_list_hidden_cmd,
+	return bgp_vty_return(vty, ret);
+}
+
+DEFUN (neighbor_advertise_map,
```
I think we can handle simpler with DEFPY and bundled with `[no$no]` form as well.

ton31337 commented on bgpd/bgpd.c:
```diff
@@ -6569,6 +6570,192 @@ int peer_unsuppress_map_unset(struct peer *peer, afi_t afi, safi_t safi)
+
+		/* Set configuration on peer-group member. */
+		filter = &member->filter[afi][safi];
+		if (filter->advmap.aname) {
```
Looks like this is duplicated code like at the beginning of this function. Can't be reused somehow nicely?

ton31337 replied:
Separate (inlined?) function or so.

maduri111 replied:
I think this is applicable for route-map and unsuppress-map as well.
introduced peer_update_rmap_filter_data() and flags for advertised-map, route-map and unsuppress-map. 
as of now modified the code for advertise-map. 
If you are OK to modify those commands also in this PR will do that otherwise will do it as separate PR.

ton31337 commented on bgpd/bgpd.h:
```diff
@@ -792,6 +802,19 @@ struct bgp_filter {
+		char *cname;
+		struct route_map *cmap;
+
+		enum advertise advertise;
```
This would sound better with something like `update_type` or so, IMO.

maduri111 replied:
enum update_type { WITHDRAW, NLRI }; is this fine ? or 
enum update_type { WITHDRAW, ADVERTISE }; ?

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,282 @@
+
+	for (dest = bgp_table_top(table); dest; dest = bgp_route_next(dest)) {
+		dest_p = bgp_dest_get_prefix(dest);
+		if (!dest_p)
```
is this gonna happen?

maduri111 replied:
may be not. will assert here

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,282 @@
+				     struct route_map *rmap)
+{
+	struct bgp_dest *dest;
+	struct attr dummy_attr;
```
can we initialize here to `{0}`?

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,282 @@
+			path.attr = &dummy_attr;
+
+			ret = route_map_apply(rmap, dest_p, RMAP_BGP, &path);
+			if (ret == RMAP_PERMITMATCH)
```
We must reset attributes it was before with `bgp_attr_flush(&dummy_attr)` if not RMAP_PERMITMATCH. Otherwise, it will lead to changed route-maps.

maduri111 replied:
If you don't mind I have a doubt in this comment.
as you said, in case of subgroup_announce_check() attributes are modified before route_map_apply() and when result is DENYMATCH, bgp_attr_flush() is called.
But here we are using the attr as it is right do we need bgp_attr_flush() ?

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,282 @@
+	assert(bgp);
+
+	bgp->t_condition_check = NULL;
+	thread_add_timer(bm->master, bgp_conditional_adv_timer, bgp,
```
I'm not sure why is this needed here?

maduri111 replied:
here we are creating thread (enable polling) which polls each peer (and for each AFI/SAFI within peer) in BGP process every 60 seconds once. this happens only once when the first advertise-map is created in BGP process.

ton31337 commented on bgpd/bgp_vty.c:
```diff
@@ -7047,6 +7047,81 @@ ALIAS_HIDDEN(no_neighbor_filter_list, no_neighbor_filter_list_hidden_cmd,
+		idx_condition_word++;
+	}
+
+	if (argv_find(argv, argc, "non-exist-map", &idx))
```
you can handle this the same way as with [no$no]

maduri111 commented:
ci:rerun

maduri111 commented:
@donaldsharp @rzalamena @ton31337 
all review comments are addressed, document and topotests are updated. logs are attached in commit messages.
Please check and provide your inputs.

ton31337 commented on bgpd/bgpd.h:
```diff
@@ -1448,6 +1480,10 @@ struct peer {
 	bool as_path_loop_detection;
 
+	/* Conditional advertisement */
+	bool advmap_config_change[AFI_MAX][SAFI_MAX];
```
why can't we put those flags under peer->af_flags?

maduri111 replied:
peer->af_flags[afi][safi] flags are updated during peer_xxxx_set/peer_xxxx_unset on peer[afi][safi].
But peer->advmap_config_change[afi][safi] is used to notify scanner function when there is a change in related ACLs/prefix-lists/route-maps/match-rules.
I have removed the other flags introduced. please find the reason in below comments.

ton31337 commented on tests/topotests/bgp_conditional_advertisement/r1/bgpd.conf:
```diff
@@ -0,0 +1,30 @@
+!
+ip prefix-list CUST seq 5 permit 10.139.224.0/20
+ip prefix-list DEFAULT seq 5 permit 0.0.0.0/0
+ip prefix-list PL1 seq 5 permit 1.1.1.1/32
```
please use https://tools.ietf.org/html/rfc5737

maduri111 replied:
using 192.0.2.0/24, 198.51.100.0/24 and 203.0.113.0/24 is Ok OR
shall I use 192.0.2.0/26 for all routers. which is preferred one?

ton31337 replied:
Only external IP addresses need to be changed.

maduri111 replied:
ok

ton31337 commented on bgpd/bgpd.h:
```diff
@@ -792,6 +810,19 @@ struct bgp_filter {
+		char *cname;
+		struct route_map *cmap;
+
+		enum update_type advertise;
```
nit: advertise -> update_type adds more readability into code itself

maduri111 replied:
changed advertise to update_type.

ton31337 commented on bgpd/bgpd.c:
```diff
@@ -6587,6 +6588,173 @@ int peer_unsuppress_map_unset(struct peer *peer, afi_t afi, safi_t safi)
+	SET_FLAG(config_flags, BGP_PEER_RMAP_SET);
+	SET_FLAG(config_flags, BGP_PEER_ADVERTISE_MAP);
+	if (condition)
+		SET_FLAG(config_flags, BGP_PEER_CONDITION_EXIST);
```
Can't this be reused with af_flags as well? My point is to reuse and keep using existing places for flags. PEER_FLAG_ADV_MAP_CONDITION_EXIST or so.

maduri111 replied:
These flags were introduced just to avoid multiple arguments to the function (which is defined to avoid duplicate code in set and unset functions) in case we want to use the same function for route-map and unsuppress-map.
Anyways, I will remove all these flags and make the function specific to advertise-map filter update.
i.e changed 
peer_update_rmap_filter_data(...., config_flags) to peer_advertise_map_filter_update(...., bool condition, bool set)

ton31337 reviewed (commented):
The overall code looks good to me, but one thing I miss is logging. Please add more debug logging for some critical conditions.

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+	if (!bgp)
+		return -1;
+
+	bgp->t_condition_check = NULL;
```
can we fix this as well?

maduri111 replied:
removed bgp->t_condition_check = NULL;

ton31337 commented on bgpd/bgpd.h:
```diff
@@ -678,9 +678,13 @@ struct bgp {
 	enum bgp_link_bw_handling lb_handling;
-
-	/* Process Queue for handling routes */
+  
```
please no whitespaces here

maduri111 replied:
sorry, i have resolved the conflict in GUI. not addressed your previous comments yet due to VM issues. will send it by EOD

maduri111 replied:
addressed this comment

maduri111 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+
+	bgp = peer->bgp;
+
+	if (!bgp)
```
addressed this comment

maduri111 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,494 @@
+		return -1;
+
+	bgp = peer->bgp;
+	if (!bgp)
```
addressed this comment

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,322 @@
+			if (ret != RMAP_PERMITMATCH)
+				bgp_attr_flush(&dummy_attr);
+			else {
+				zlog_debug(
```
log only if BGP DEBUG is enabled

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,322 @@
+		}
+	}
+
+	zlog_debug("%s: Condition map routes not present in BGP table",
```
here as well, we need log if explicit logging is enabled

maduri111 replied:
do you mean to add [no] debug bgp conditional-advertisement ?

ton31337 replied:
no, I mean at least if `debug bgp updates` enabled.

maduri111 replied:
ok i will add the check if (BGP_DEBUG(update, UPDATE_OUT))

maduri111 replied:
addressed this comment

ton31337 commented on bgpd/bgp_conditional_adv.c:
```diff
@@ -0,0 +1,335 @@
+
+			if (BGP_DEBUG(update, UPDATE_OUT))
+				zlog_debug(
+					"%s: %s - advmap_table_change %d advmap_config_change[%d][%d] %d ",
```
[%d][%d] seems not very handy, I think converting to a string would be better.

maduri111 replied:
Replaced afi/safi with string and printing the debug message only if the variable is set.

ton31337 commented on bgpd/bgpd.c:
```diff
@@ -6585,6 +6586,175 @@ int peer_unsuppress_map_unset(struct peer *peer, afi_t afi, safi_t safi)
+
+	/* advertise-map is not configured */
+	if (!peer->filter[afi][safi].advmap.aname) {
+		if (BGP_DEBUG(update, UPDATE_OUT))
```
I doubt this is handy at all here.

maduri111 replied:
Ok. Removed this debug message.

maduri111 commented:
ci:rerun

ton31337 reviewed (approved):
LGTM, @maduri111 thanks.

maduri111 commented:
> 
> 
> LGTM, @maduri111 thanks.

@ton31337 Thanks for reviewing and approving the changes.

rzalamena commented on tests/topotests/bgp_conditional_advertisement/test_bgp_conditional_advertisement.py:
```diff
@@ -0,0 +1,961 @@
+# Part of NetDEF Topology Tests
+#
+# Copyright (c) 2020 by
+# Network Device Education Foundation, Inc. ("NetDEF")
```
Wrong copyright?

maduri111 replied:
> 
> 
> Wrong copyright?

@rzalamena I thought it should not be changed. will change it. Thanks for reviewing and approving the code changes

rzalamena reviewed (approved):
Looks good to me, just two nits in topotest:

- Author forgot to change copyright owner
- The tests could be split in more functions to help finding a failure, but I guess it is also easy to search for `TCXX` (where XX is a number).
